Task library

iryna-pantsialeyeva opened the pull request:
Please review my PR

iryna-pantsialeyeva committed:
add task8.Part2: Classes and Objects tasks from book Part 2 p.25

iryna-pantsialeyeva committed:
add an obligatory part of task "Library"

DimaLamaka commented on src/library/Library.java:
```diff
@@ -0,0 +1,85 @@
+
+            if (bookFromLibrary.getId() == book.getId()) {
+                isBookInLibrary = true;
+                break;
```
лучше убрать переменную isBookInLibrary  и в этом if если нашли совпадение, распечатали сообщение и вышли сразу же из метода через return

RomanKashenok replied:
It's better to keep possibility to add multiple similar books (they should differs by ID). Every library has multiple similar books.

iryna-pantsialeyeva replied:
Dear Dima, thank you so much for your recommendation. Its realisation is in the commented code (lines 20-27).
Dear Roman, many thanks for your advice. Done (lines 14-16)

DimaLamaka commented on src/library/Library.java:
```diff
@@ -0,0 +1,85 @@
+
+public class Library {
+
+    private List<Book> library;
```
books

iryna-pantsialeyeva replied:
Thank you. Done

DimaLamaka commented on src/library/Library.java:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    public List<Book> getLibrary() { // получить список всех книг
```
getBooks()

iryna-pantsialeyeva replied:
Thank you once again. Done

DimaLamaka commented on src/library/Library.java:
```diff
@@ -0,0 +1,85 @@
+            int index = library.indexOf(getBookById(id));
+            library.remove(index);
+        }
+    }
```
зачем два раза искать книгу? вынеси выше получение книги по id, и если есть, то у листа  метод remove() принимает объект, который ты уже и нашла выше

iryna-pantsialeyeva replied:
Many thanks. Done

iryna-pantsialeyeva replied:
Many thanks. Done

DimaLamaka commented on src/task8/booktasks/atm/Atm.java:
```diff
@@ -0,0 +1,78 @@
+        setUsd100(getUsd100() + usd100);
+        setUsd50(getUsd50() + usd50);
+        setUsd20(getUsd20() + usd20);
+        return getAmountInAtm();
```
зачем использовать сеттеры и геттеры, когда у тебя есть к ним прямой доступ внутри класса ?

DimaLamaka commented on src/task8/booktasks/atm/Atm.java:
```diff
@@ -0,0 +1,78 @@
+            this.usd20 -= ((neededMoney % 100) % 50) / 20;
+            return true;
+        }
+        return false;
```
неправильная логика метода

DimaLamaka commented on src/task8/booktasks/timeline/Timeline.java:
```diff
@@ -0,0 +1,58 @@
+    private int min;
+    private int sec;
+
+    public Timeline(int hour, int min, int sec) {
```
желательно было бы сделать проверку на корректность данных, которые тебе передаются

DimaLamaka commented on src/task8/booktasks/timeline/Timeline.java:
```diff
@@ -0,0 +1,58 @@
+
+    public int compareTo(Timeline timeline) {
+        int index = this.calculateSeconds() - timeline.calculateSeconds();
+        return index;
```
можно в одну строку return this.calculateSeconds() - timeline.calculateSeconds();

DimaLamaka commented on src/task8/booktasks/timeline/Timeline.java:
```diff
@@ -0,0 +1,58 @@
+            System.out.println("timeline1 = timeline2");
+        }
+    }
+}
```
метод print по названию, должен просто распечатать сам интервал, а не сравнение

RomanKashenok commented on src/library/Book.java:
```diff
@@ -0,0 +1,68 @@
+    private String genre;
+
+    public Book(int id, String title, String genre) {
+        this.id = id;
```
let's generate ID by ourself (every new book should have ID greater than previous book on 1)

iryna-pantsialeyeva replied:
Done

RomanKashenok commented on src/library/Book.java:
```diff
@@ -0,0 +1,68 @@
+        return id;
+    }
+
+    public void setId(int id) {
```
we should avoid possibility to update IDs

iryna-pantsialeyeva replied:
Done

RomanKashenok commented on src/library/Book.java:
```diff
@@ -0,0 +1,68 @@
+    @Override
+    public int compareTo(Book o) {
+        return this.getTitle().toLowerCase().compareTo(o.getTitle().toLowerCase());
+    }
```
Let's keep comparing strategy not hardcoded and depending on users chose. remove it

iryna-pantsialeyeva replied:
Done. Comparable and method compareTo were removed. Comparator was used.

RomanKashenok commented on src/library/Library.java:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    public void reversePrinter() {
```
what is the purpose of this method? o_O

iryna-pantsialeyeva replied:
The purpose is " • получаем список книг из библиотеки, сортируем по порядку добавления (последние добавленные в начале) и выводим в консоль. ". Because we use ArrayList the order of the books corresponds to the order in which they were added. The method is replaced to main as requested in the task.

RomanKashenok commented on src/library/Library.java:
```diff
@@ -0,0 +1,85 @@
+        }
+    }
+
+    public void printer() {
```
and purpose of this method too...

iryna-pantsialeyeva replied:
The purpose was "• получаем список книг и выводим в консоль "
Fixed. A separate method printer was created in the class Library.

RomanKashenok commented on src/library/Runner.java:
```diff
@@ -0,0 +1,31 @@
+        library.addBook(book3);
+        library.addBook(book4);
+
+        Collections.sort(library.getLibrary());
```
I'd suggest to add some comparators to Library class and pass them into sort() method with list of books. This way we keep multiple possibilities to sort library depending on users choise

iryna-pantsialeyeva replied:
Many thanks for your recommendation. Done

RomanKashenok reviewed (changes requested):
reviewed only Library task (as in title of MR)

iryna-pantsialeyeva committed:
add the corrected task "Library"

iryna-pantsialeyeva committed:
add the corrected task "Library" according to comments in the PR about sorting with usage of Comparator

RomanKashenok commented on src/library/Book.java:
```diff
@@ -47,21 +47,11 @@ public String toString() {
     }
-
-    @Override
-    public int hashCode() {
```
nonono. hashCode and equals functions should come together. revert this rows please

iryna-pantsialeyeva replied:
Done

RomanKashenok commented on src/library/Library.java:
```diff
@@ -27,10 +34,6 @@ public void addBook(Book book) {
 //       books.add(book);
```
remove commented blocks please

iryna-pantsialeyeva replied:
Done

RomanKashenok commented on src/library/Library.java:
```diff
@@ -75,4 +78,41 @@ public int getLibrarySize() {
         return books.get(i);
     }
+
+    public List<Book> getSortedBooks(String sortingOrder, String sortingType) {
```
yeah ))) almost cool.
You can move sorting orders to enum with values 1 and 0 for example. And use it instead of comparing Strings. But this way it will works too.

RomanKashenok reviewed (approved):
Верни hashCode, убери комменты и мержи

iryna-pantsialeyeva committed:
add the corrected task "Library"
according to comments in the PR about
returning hashCode method
removing commented method add() in Library
